sklearn/metrics/cluster/unsupervised.py

```diff
@@ -255,3 +261,70 @@ def calinski_harabaz_score(X, labels):
     return (1. if intra_disp == 0. else
             extra_disp * (n_samples - n_labels) /
             (intra_disp * (n_labels - 1.)))
+
+
+def prediction_strength_score(labels_train, labels_test):
+    """Compute the prediction strength score.
+
+    For each test cluster, we compute the proportion of observation pairs
+    in that cluster that are also assigned to the same cluster by the
+    training set centroids. The prediction strength is the minimum of this
+    quantity over the k test clusters.
+
+    The best value is 1.0 (if the assignments of `labels_train` and
+    `labels_test` are identical) and the worst value is 0 (if all samples of
+    one cluster of `labels_test` are not co-members of some cluster in
+    `labels_train`).
+
+    Parameters
+    ----------
+    labels_train : array-like, shape (``n_test_samples``,)
+        Predicted labels for each sample in the the test data
+        based on clusters derived from independent training data.
+
+    labels_test : array-like, shape (``n_test_samples``,)
+        Predicted labels for each sample in the test data
+        based on clusters derived from the same data.
+
+    Returns
+    -------
+    score : float
+        The resulting prediction strength score.
+
+    References
+    ----------
+    .. [1] `Robert Tibshirani and Guenther Walther (2005). "Cluster Validation
+    by Prediction Strength". Journal of Computational and Graphical Statistics,
+    14(3), 511-528. <http://doi.org/10.1198/106186005X59243>_`
+    """
+    check_consistent_length(labels_train, labels_test)
+
+    labels_train = check_array(labels_train, dtype=np.int32, ensure_2d=False,
+                               warn_on_dtype=True)
+    labels_test = check_array(labels_test, dtype=np.int32, ensure_2d=False,
+                              warn_on_dtype=True)
+
+    clusters = set(chain(labels_train, labels_test))
+    n_clusters = len(clusters)
+    if n_clusters == 1:
+        return 1.0  # by definition
+
+    strength = 1.0
+    for k in clusters:
+        # samples assigned to k-th cluster based on test data
+        samples_test_k = np.flatnonzero(labels_test == k)
+        cluster_test_size = samples_test_k.shape[0]
+
+        if cluster_test_size < 2:
+            continue
+
+        matches = 0
+        for i, j in permutations(xrange(cluster_test_size), 2):
```

Comment:
I don't think `contingency_matrix` would work here, because we want to check whether pairs of samples in `samples_test_k` are assigned the same cluster in `labels_train` too. Most importantly, the actual label of the cluster is irrelevant. Assume a pair `(x,y)` of samples from `labels_test` with label `k=0`, then check whether the label of `x` equals the label of `y` according to `labels_train`, if `k != 0` in `labels_train`, that's okay. Please see the unit tests for examples.